### PR TITLE
Add getContent API

### PR DESCRIPTION
**Pull request by ShanmathiMayuramKrithivasan**
## Description

This PR introduces `sharing.history.getContent` API to fetch content shared in a Teams context.

### Main changes in the PR:

1. Added a new sub-namespace `history` in `sharing` namespace.
2. Added interface `IContentResponse` for the content details params that would be returned.
3. Added support for 'getContent' API.
4. Updated `sharing.spec.ts` with unit tests for this new API.

### Unit Tests added:
Yes

### End-to-end tests added:
Existing Teams repo end-to-end test to be updated.

### Change file added:
Yes

### Next/remaining steps:

- [ ] Teams changes
- [ ] Hosts SDK changes

### CODE REVIEW

**Comment by vikramtha** [comment]
Hello, I am the PM assigned to review this. I have just reached out to Shanmathi Mayuram Krithivasan on teams asking about timelines

**Comment by AE-MS** [comment]
Please make sure you add a changefile using the guidelines in the [contributor's guide](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball). The build will not pass until it is present. Thanks!

File: packages/teams-js/src/public/sharing.ts [hunk 1]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+      /** Id of the app where the content was shared from */
+      appId: string;
+      /** Title of the shared content */
+      title: string;
+      /** Reference of the shared content */
+      contentReference: string;
+      /** Id of the thread where the content was shared */
+      threadId: string;
+      /** Id of the user who shared the content */
+      author: string;
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:219]: What format is this in?

File: packages/teams-js/src/public/sharing.ts [hunk 2]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+      /** Type of the shared content */
+      contentType: string;
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:221]: Can this really be any string? Or is there a list of possible content types? Is it a MIME type?

File: packages/teams-js/src/public/sharing.ts [hunk 3]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+     */
+    export interface IContentResponse {
+      /** Id of the app where the content was shared from */
+      appId: string;
+      /** Title of the shared content */
+      title: string;
+      /** Reference of the shared content */
+      contentReference: string;
+      /** Id of the thread where the content was shared */
+      threadId: string;
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:217]: Can thread IDs really be any string or is there a format to them? Is "Shanmathi" a valid thread ID?

File: packages/teams-js/src/public/sharing.ts [hunk 4]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+
+    /**
+     * Get the list of content shared in a context
+     * @param callback - Callback contains 2 parameters, `error` and `contentDetails`.
+     * `error` can either contain an error of type `SdkError`, in case of an error, or null when fetch is successful.
+     * `contentDetails` will be the list of contents {@link IContentResponse} that was shared in the context, or null when the request fails.
+     *
+     * @beta
+     */
+    export function getContent(callback: getContentCallbackFunctionType): void {
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:242]: >callback: getContentCallbackFunctionType [](http://example.com/codeflow?start=31&length=40)

Please return a `Promise` instead of using a callback. Thanks!

File: packages/teams-js/src/public/sharing.ts [hunk 5]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+      if (!callback) {
+        throw new Error('[get content] Callback cannot be null');
+      }
+      ensureInitialized(runtime, FrameContexts.sidePanel, FrameContexts.meetingStage);
+      if (!isSupported()) {
+        throw errorNotSupportedOnPlatform;
+      }
+      sendMessageToParent(
+        getApiVersionTag(sharingTelemetryVersionNumber_v1, ApiName.Sharing_History_GetContent),
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:251]: >sharingTelemetryVersionNumber_v1 [](http://example.com/codeflow?start=25&length=32)

I actually don't know how these version numbers are supposed to be set. @KangxuanYe what's the right constant to use for new functions? Is it v1 like this?

File: packages/teams-js/src/public/sharing.ts (hunk 1 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:219]: author is the userID here and will be GUID.

File: packages/teams-js/src/public/sharing.ts [hunk 6]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+     * @throws Error if {@linkcode app.initialize} has not successfully completed
+     *
+     * @beta
+     */
+    export function isSupported(): boolean {
+      return ensureInitialized(runtime) && runtime.supports.sharing
+        ? runtime.supports.sharing.history
+          ? true
+          : false
+        : false;
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:270]: This is a bit difficult to follow when written like this. Consider instead something like:
`return ensureInitialized(runtime) && runtime.supports.sharing?.history !== undefined`

File: packages/teams-js/src/public/sharing.ts (hunk 2 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:221]: These are string types. Right now on client the contentType is "ShareToStage" for this scenario.

File: packages/teams-js/src/public/sharing.ts (hunk 3 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:217]: No, threadIds are GUIDs that are returned from Teams side.

File: packages/teams-js/src/public/sharing.ts [hunk 7]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
   }
+
+  /**
+   * Namespace to get the list of content shared in a context
+   *
+   * @beta
+   */
+  export namespace history {
+    /**
+     * Represents IContentResponse parameters.
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:205]: >Represents IContentResponse parameters. [](http://example.com/codeflow?start=7&length=39)

Consider:

"Represents the data returned when calling {@link sharing.history.getContent}"

File: packages/teams-js/src/public/sharing.ts [hunk 8]
```diff
@@ -194,4 +194,80 @@ export namespace sharing {
...
+     *
+     * @beta
+     */
+    export type getContentCallbackFunctionType = (
+      error: SdkError | null,
+      contentDetails: IContentResponse[] | null,
+    ) => void;
+
+    /**
+     * Get the list of content shared in a context
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:235]: >shared in a context [](http://example.com/codeflow?start=31&length=19)

Can you elaborate on what "shared in a context" means? I'm not sure that developers reading this would know how to interpret that phrase.

File: packages/teams-js/src/internal/telemetry.ts [hunk 9]
```diff
@@ -290,6 +290,7 @@ export const enum ApiName {
   Settings_Save_Success = 'settings.save.success',
   Settings_SetSettings = 'settings.setSettings',
   Settings_SetValidityState = 'settings.setValidityState',
+  Sharing_History_GetContent = 'sharing.getContent',
```
Comment (AE-MS) [inline packages/teams-js/src/internal/telemetry.ts:293]: >sharing.getContent [](http://example.com/codeflow?start=32&length=18)

Should this be `sharing.history.getContent` ?

**Review by AE-MS — CHANGES_REQUESTED** [review]
:clock1:

**Comment by AE-MS** [comment]
I haven't looked at the new unit tests yet since one of the changes I requested will change the function signature of `getContent` -- I will look at them on the next iteration. Thanks for considering my feedback!

**Comment by AE-MS** [comment]
Please don't forget to response to the earlier comment from the github action bot about the runtime change you made. Thanks!

**Comment by AE-MS** [comment]
Your build is failing because it thinks you are missing a changefile. I see you do have a changefile in here, so I'm not totally sure why it is complaining. Did you create the file using the `pnpm changefile` command? One thing I notice is that you list the change type as `beta` which isn't one of the four valid types. Perhaps that is what is causing the problem?
See more detail in the [contributor's guide](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

File: packages/teams-js/src/public/sharing.ts (hunk 1 above)
Comment (AE-MS) [reply packages/teams-js/src/public/sharing.ts:219]: Cool; can you use a type that only allows for UUIDs then, rather than `string`? Or if you really must use `string`, can you validate that any values you receive are valid UUIDs?

Please also document the format. Thanks!

File: packages/teams-js/src/public/sharing.ts (hunk 3 above)
Comment (AE-MS) [reply packages/teams-js/src/public/sharing.ts:217]: Same comment as for `author`, then. Thanks!

File: packages/teams-js/src/public/sharing.ts (hunk 2 above)
Comment (AE-MS) [reply packages/teams-js/src/public/sharing.ts:221]: So there's only one possible value? Can you document the valid values so that developers know what to expect?

Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:227]: Added validation and updated documentation.

File: packages/teams-js/src/public/sharing.ts (hunk 3 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:221]: Added validation and updated documentation.

File: packages/teams-js/src/public/sharing.ts (hunk 4 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated to return a promise

File: packages/teams-js/src/public/sharing.ts (hunk 6 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated

File: packages/teams-js/src/public/sharing.ts (hunk 8 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated comment

File: packages/teams-js/src/internal/telemetry.ts (hunk 9 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/internal/telemetry.ts]: yes updated

**Comment by ShanmathiMayuramKrithivasan** [comment]
> This pull request contains changes to the runtime.ts file. If you, as the author of this PR, have made changes to the Runtime interface please review RUNTIME.md to determine if a new runtime version is required. Please reply to this comment stating what changes, if any, were made to the runtime object and whether a new runtime version was required.


Based on the RUNTIME.md file, in this PR we have added a new optional property or supported capability. So will not require a new version.

**Comment by ShanmathiMayuramKrithivasan** [comment]
> Your build is failing because it thinks you are missing a changefile. I see you do have a changefile in here, so I'm not totally sure why it is complaining. Did you create the file using the `pnpm changefile` command? One thing I notice is that you list the change type as `beta` which isn't one of the four valid types. Perhaps that is what is causing the problem? See more detail in the [contributor's guide](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

Created a new change file. This should fix the build errors.

File: packages/teams-js/src/public/sharing.ts [hunk 10]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
...
+      author: string;
+      /** Type of the shared content.
+       * For sharing to Teams stage scenarios, this value would be ShareToStage
+       */
+      contentType: string;
+    }
+
+    /**
+     * Get the list of content shared in a Teams meeting
+     *
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:232]: >* [](http://example.com/codeflow?start=5&length=1)

Please document the return type using the `@returns` doc format. You can find some other examples in this file and elsewhere.

File: packages/teams-js/src/public/sharing.ts [hunk 11]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
...
+    export function getContent(): Promise<IContentResponse[]> {
+      ensureInitialized(runtime, FrameContexts.sidePanel, FrameContexts.meetingStage);
+      if (!isSupported()) {
+        throw errorNotSupportedOnPlatform;
+      }
+
+      return sendAndHandleSdkError(
+        getApiVersionTag(sharingTelemetryVersionNumber_v1, ApiName.Sharing_History_GetContent),
+        'sharing.history.getContent',
+      ).then((contentDetails: IContentResponse[]) => {
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:244]: >then [](http://example.com/codeflow?start=8&length=4)

There's nothing wrong with how you wrote this but it would likely be easier to read if you use `await` instead of `.then`. Your call, though.

File: packages/teams-js/src/public/sharing.ts [hunk 12]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
...
+        contentDetails.map((contentDetails) => {
+          validateUuid(contentDetails.author);
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:246]: >validateUuid [](http://example.com/codeflow?start=10&length=12)

I looked into the implementation of these functions and they seem mostly correct but they weirdly "allow" for the input parameter to be `undefined` or `null` without complaining. This seems wrong to me (in general, not specifically for your scenario), since `undefined` and `null` are both invalid UUIDs.

To be sure your validation is comprehensive, can I suggest you check here to ensure they aren't `null` or `undefined`? I know it's lame that the shared function that someone wrote doesn't do the right thing and that you can't just use that shared function, but I'd rather you have the correct validation and not suffer because of a suboptimal shared function. You could even help the whole codebase by adding a new function next to `validateUuid` in the shared location that doesn't tolerate `undefined` or `null`, e.g. `validateDefinedUuid` or something. Whatever works for you.

(unless it is actually valid for these two values to be `null` or `undefined` for you, in which case you should update their types on the `IContentResponse` interface)

File: packages/teams-js/src/internal/utils.ts [hunk 13]
```diff
@@ -400,3 +400,22 @@ export function ssrSafeWindow(): Window {
...
+ * @param id The ID to validate against the UUID format
+ * @throws Error if ID is not a valid UUID
+ *
+ * @internal
+ * Limited to Microsoft-internal use
+ */
+export function validateUuid(id: string | undefined | null): void {
+  if (id === undefined || id === null) {
+    return;
+  }
```
Comment (AE-MS) [inline packages/teams-js/src/internal/utils.ts:414]: Is it correct that `undefined` and `null` are "valid" UUIDs? Aren't those invalid?

File: packages/teams-js/test/internal/utils.spec.ts [hunk 14]
```diff
@@ -185,4 +186,22 @@ describe('utils', () => {
       }
     });
   });
+
+  describe('validateUuid', () => {
```
Comment (AE-MS) [inline packages/teams-js/test/internal/utils.spec.ts:190]: >validateUuid [](http://example.com/codeflow?start=12&length=12)

Can you please add a unit test for the empty string case (it should be invalid)?

Comment (AE-MS) [reply packages/teams-js/test/internal/utils.spec.ts:190]: (the answer to this depends on the answer to my other comment in `utils.ts` about why `undefined` and `null` are considered valid. There should also be unit tests here for those cases)

File: packages/teams-js/src/public/sharing.ts [hunk 15]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
...
+      /** Title of the shared content */
+      title: string;
+      /** Reference of the shared content */
+      contentReference: string;
+      /** Id of the thread where the content was shared. This is a UUID */
+      threadId: string;
+      /** Id of the user who shared the content. This is a UUID */
+      author: string;
+      /** Type of the shared content.
+       * For sharing to Teams stage scenarios, this value would be ShareToStage
```
Comment (AE-MS) [inline packages/teams-js/src/public/sharing.ts:225]: >ShareToStage [](http://example.com/codeflow?start=67&length=12)

Please surround this with backticks, e.g.,

For sharing to Teams stage scenarios, this value would be `ShareToStage`

And also add:
Other `contentType` values will be added here over time
(or some other text that indicates that there are currently no other known values but it will expand in the future)

File: packages/teams-js/test/public/sharing.spec.ts [hunk 16]
```diff
@@ -694,4 +696,104 @@ describe('sharing_v2', () => {
         }
       });
   });
+
+  describe('Testing getContent', () => {
```
Comment (AE-MS) [inline packages/teams-js/test/public/sharing.spec.ts:700]: >describe [](http://example.com/codeflow?start=2&length=8)

Can you please add two tests:
- one that ensures that an error is thrown if the authorid is not a valid uuid
- one that ensures that an error is thrown if the threadid is not a valid uuid

**Comment by AE-MS** [comment]
I didn't see any changes made to the test app to validate this new capability, and similarly there are no new E2E tests added. I know that you can't actually run those E2E scenarios or tests until the corresponding change is made in the host SDK. Please ensure you come back and add them after you make that host SDK change. Thanks!

File: packages/teams-js/src/public/sharing.ts [hunk 17]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
...
+       */
+      contentType: string;
+    }
+
+    /**
+     * Get the list of content shared in a Teams meeting
```
[the remark this answers is not in the document]
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts:228]: Updated

File: packages/teams-js/src/public/sharing.ts (hunk 11 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated to use `await`.

File: packages/teams-js/src/internal/utils.ts (hunk 13 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/internal/utils.ts]: Updated to make all these cases invalid.

File: packages/teams-js/test/internal/utils.spec.ts (hunk 14 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/test/internal/utils.spec.ts:226]: Added tests for `undefined`, `null` and empty string.

File: packages/teams-js/src/public/sharing.ts [hunk 18]
```diff
@@ -194,4 +198,68 @@ export namespace sharing {
   export function isSupported(): boolean {
     return ensureInitialized(runtime) && runtime.supports.sharing ? true : false;
   }
+
+  /**
+   * Namespace to get the list of content shared in a context
```
[the remark this answers is not in the document]
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated description.

File: packages/teams-js/src/public/sharing.ts (hunk 15 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/src/public/sharing.ts]: Updated and added the comments.

File: packages/teams-js/test/public/sharing.spec.ts (hunk 16 above)
Comment (ShanmathiMayuramKrithivasan) [reply packages/teams-js/test/public/sharing.spec.ts:700]: Updated tests for validating valid uuid,

**Comment by ShanmathiMayuramKrithivasan** [comment]
> I didn't see any changes made to the test app to validate this new capability, and similarly there are no new E2E tests added. I know that you can't actually run those E2E scenarios or tests until the corresponding change is made in the host SDK. Please ensure you come back and add them after you make that host SDK change. Thanks!

Yes will add E2E tests once the host SDK changes are in.

**Review by AE-MS — DISMISSED** [review]

File: packages/teams-js/src/public/sharing.ts (hunk 5 above)
Comment (KangxuanYe) [reply packages/teams-js/src/public/sharing.ts:243]: This is V2 rather than V1. Only API exists in Teams-JS library v1 and does NOT exist in Teams-JS library v2 will be labelled as V1. Vice versa.

This new API is designed to be in Teams-JS library V2 so it should fix here from v1 to v2.

**Review by KangxuanYe — CHANGES_REQUESTED** [review]
Need to fix telemetry version

File: packages/teams-js/src/public/sharing.ts [hunk 19]
```diff
@@ -194,4 +194,73 @@ export namespace sharing {
...
+     * @beta
+     */
+    export async function getContent(): Promise<IContentResponse[]> {
+      ensureInitialized(runtime, FrameContexts.sidePanel, FrameContexts.meetingStage);
+      if (!isSupported()) {
+        throw errorNotSupportedOnPlatform;
+      }
+
+      const contentDetails: IContentResponse[] = await sendAndHandleSdkError(
+        getApiVersionTag(sharingTelemetryVersionNumber_v1, ApiName.Sharing_History_GetContent),
```
Comment (KangxuanYe) [inline packages/teams-js/src/public/sharing.ts:243]:
```suggestion
        getApiVersionTag(sharingTelemetryVersionNumber_v2, ApiName.Sharing_History_GetContent),
```

**Review by KangxuanYe — CHANGES_REQUESTED** [review]
Proposed a suggestion.

**Review by KangxuanYe — APPROVED** [review]

**Review by AE-MS — APPROVED** [review]
